crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1553,12 +1553,14 @@ impl<'db> TypeInferenceBuilder<'db> {
             parenthesized: _,
         } = tuple;
 
-        for elt in elts {
-            self.infer_expression(elt);
-        }
+        let element_types = elts
+            .iter()
+            .map(|elt| self.infer_expression(elt))
+            .collect::<Vec<_>>();
 
-        // TODO generic
-        builtins_symbol_ty(self.db, "tuple").to_instance()
+        // TODO: Currently, we support only tuple literals like `(1, 2)` or `(x, y)` but we still
+        // need to support generic tuple types like `tuple[int, str]` and `tuple[int, ...]`.
```

Comment:
I don't understand the reference to `tuple[int, str]` here -- it seems to me that this representation can already represent that type just fine. The only one it can't represent is `tuple[int, ...]`. (We don't yet ever infer `tuple[int, str]`, e.g. from an annotation, but that just requires adding a bit of handling to `infer_annotation_expression`, it doesn't require any change to the code added in this PR.)

I also don't think this TODO belongs at this spot in the code. This code (for inferring a tuple type from a literal) is already fully correct, there's nothing more to be done here. If the TODO belongs anywhere it's in `types.rs`, in the list of `Type` variants, as a TODO that we don't yet have a representation for variable-length homogenous tuples (`tuple[T, ...]`)

Reply:
That makes sense, thanks. I'll update the TODO and it's location